File: QuantaraCore/Sources/QuantaraCore/Engine/Statistics.swift
import Foundation

/// Statistiques descriptives sur des `Decimal`.
///
/// Réécrites plutôt qu'empruntées à `Double` : convertir des montants en virgule
/// flottante pour calculer une médiane réintroduit précisément l'imprécision que
/// `Decimal` sert à éviter.
public enum DecimalStatistics {

    public static func sum(_ values: [Decimal]) -> Decimal {
        values.reduce(Decimal.zero, +)
    }

    public static func mean(_ values: [Decimal]) -> Decimal? {
        guard !values.isEmpty else { return nil }
        return sum(values) / Decimal(values.count)
    }

    /// Médiane. Préférée à la moyenne pour les revenus irréguliers et les dépenses par
    /// catégorie : un seul mois exceptionnel ne doit pas déplacer la référence.
    public static func median(_ values: [Decimal]) -> Decimal? {
        guard !values.isEmpty else { return nil }
        let sorted = values.sorted()
        let middle = sorted.count / 2
        if sorted.count % 2 == 1 { return sorted[middle] }
        return (sorted[middle - 1] + sorted[middle]) / 2
    }

    /// Percentile par interpolation linéaire. `fraction` dans [0, 1].
    public static func percentile(_ values: [Decimal], fraction: Decimal) -> Decimal? {
        guard !values.isEmpty else { return nil }
        let sorted = values.sorted()
        if sorted.count == 1 { return sorted[0] }
        let clamped = min(max(fraction, 0), 1)
        let position = clamped * Decimal(sorted.count - 1)
        var lowerDecimal = Decimal()
        var mutablePosition = position
        NSDecimalRound(&lowerDecimal, &mutablePosition, 0, .down)
        let lowerIndex = NSDecimalNumber(decimal: lowerDecimal).intValue
        let upperIndex = min(lowerIndex + 1, sorted.count - 1)
        let weight = position - lowerDecimal
        return sorted[lowerIndex] + (sorted[upperIndex] - sorted[lowerIndex]) * weight
    }

    /// Écart-type de population. Utilisé pour repérer les dépenses inhabituelles.
    public static func standardDeviation(_ values: [Decimal]) -> Decimal? {
        guard values.count > 1, let average = mean(values) else { return nil }
        let variance = sum(values.map { ($0 - average) * ($0 - average) }) / Decimal(values.count)
        return squareRoot(variance)
    }

    /// Racine carrée par la méthode de Héron. `Decimal` n'expose pas `sqrt` ;
    /// vingt itérations suffisent très largement à la précision d'affichage.
    public static func squareRoot(_ value: Decimal) -> Decimal {
        guard value > .zero else { return .zero }
        var estimate = value / 2
        if estimate == .zero { estimate = value }
        for _ in 0..<20 {
            guard estimate != .zero else { break }
            let next = (estimate + value / estimate) / 2
            if next == estimate { break }
            estimate = next
        }
        return estimate
    }

    /// Élévation à une puissance entière positive par exponentiation rapide.
    /// Base des calculs d'intérêts composés.
    public static func power(_ base: Decimal, _ exponent: Int) -> Decimal {
        guard exponent >= 0 else {
            let positive = power(base, -exponent)
            return positive == .zero ? .zero : 1 / positive
        }
        var result = Decimal(1)
        var multiplier = base
        var remaining = exponent
        while remaining > 0 {
            if remaining & 1 == 1 { result *= multiplier }
            multiplier *= multiplier
            remaining >>= 1
        }
        return result
    }
}

extension Array where Element == Money {
    /// Médiane d'une série de montants.
    public func median(currency: Currency) -> Money? {
        guard let value = DecimalStatistics.median(map(\.amount)) else { return nil }
        return Money(value, currency)
    }

    public func mean(currency: Currency) -> Money? {
        guard let value = DecimalStatistics.mean(map(\.amount)) else { return nil }
        return Money(value, currency)
    }
}

File: desktop/src/core/engine/statistics.ts
import { Money, divRound } from '../money';
import type { Currency } from '../money';

/**
 * Statistiques sur des montants.
 *
 * La médiane est préférée à la moyenne partout où un mois exceptionnel ne doit pas
 * fausser la référence : un remboursement de 3 000 € déplace la moyenne, pas la médiane.
 */
export const Statistics = {
  sum(values: readonly Money[], currency: Currency): Money {
    return Money.sum(values, currency);
  },

  mean(values: readonly Money[], currency: Currency): Money {
    if (values.length === 0) return Money.zero(currency);
    return Money.sum(values, currency).dividedBy(BigInt(values.length));
  },

  median(values: readonly Money[], currency: Currency): Money {
    if (values.length === 0) return Money.zero(currency);
    const sorted = [...values].sort((a, b) => (a.micros < b.micros ? -1 : a.micros > b.micros ? 1 : 0));
    const middle = Math.floor(sorted.length / 2);
    if (sorted.length % 2 === 1) return sorted[middle]!;
    const low = sorted[middle - 1]!;
    const high = sorted[middle]!;
    return Money.fromMicros(divRound(low.micros + high.micros, 2n), currency);
  },

  /** Percentile par interpolation linéaire, `ratio` entre 0 et 1. */
  percentile(values: readonly Money[], ratio: number, currency: Currency): Money {
    if (values.length === 0) return Money.zero(currency);
    const sorted = [...values].sort((a, b) => (a.micros < b.micros ? -1 : a.micros > b.micros ? 1 : 0));
    const position = Math.min(Math.max(ratio, 0), 1) * (sorted.length - 1);
    const lower = Math.floor(position);
    const upper = Math.ceil(position);
    if (lower === upper) return sorted[lower]!;
    const weight = position - lower;
    const delta = sorted[upper]!.micros - sorted[lower]!.micros;
    return Money.fromMicros(sorted[lower]!.micros + BigInt(Math.round(Number(delta) * weight)), currency);
  },

  standardDeviation(values: readonly Money[], currency: Currency): Money {
    if (values.length < 2) return Money.zero(currency);
    const average = Statistics.mean(values, currency).units;
    const variance =
      values.reduce((total, value) => total + (value.units - average) ** 2, 0) / (values.length - 1);
    return Money.of(Math.sqrt(variance), currency);
  },

  /** Coefficient de variation : mesure l'irrégularité d'un revenu, indépendamment de
   *  son niveau. `null` si la moyenne est nulle. */
  coefficientOfVariation(values: readonly Money[], currency: Currency): number | null {
    const average = Statistics.mean(values, currency);
    if (average.isZero) return null;
    return Statistics.standardDeviation(values, currency).units / Math.abs(average.units);
  },
};
